File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/di/DatabaseModule.java
package com.example.postscommentsapp.di;

import android.content.Context;

import androidx.room.Room;

import com.example.postscommentsapp.data.local.dao.CommentDao;
import com.example.postscommentsapp.data.local.dao.PostDao;
import com.example.postscommentsapp.data.local.db.AppDatabase;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class DatabaseModule {

    @Provides
    @Singleton
    public AppDatabase provideDatabase(@ApplicationContext Context context) {
        return Room.databaseBuilder(
                        context,
                        AppDatabase.class,
                        "posts_comments_db"
                ).fallbackToDestructiveMigration()
                .build();
    }

    @Provides
    @Singleton
    public PostDao providePostDao(AppDatabase db) {
        return db.postDao();
    }

    @Provides
    @Singleton
    public CommentDao provideCommentDao(AppDatabase db) {
        return db.commentDao();
    }
}

File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/di/DatabaseModule.kt
package com.example.postscommentsapp.di

import android.content.Context
import androidx.room.Room
import com.example.postscommentsapp.data.local.dao.CommentDao
import com.example.postscommentsapp.data.local.dao.PostDao
import com.example.postscommentsapp.data.local.db.AppDatabase
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module //Instruction to create objects
@InstallIn(SingletonComponent::class) //Important, only one db to all app.
object DatabaseModule {

    @Provides //It's a warning to say to the phone for create a database
    @Singleton //Only one db on all app
    fun provideDatabase(@ApplicationContext context: Context): AppDatabase =
        Room.databaseBuilder(
            context,
            AppDatabase::class.java,
            "posts_comments_db"
        ).fallbackToDestructiveMigration() //If db changes tha old one is destroyed
            .build()

    @Provides
    @Singleton
    fun providePostDao(db: AppDatabase): PostDao = db.postDao()

    @Provides
    @Singleton
    fun provideCommentDao(db: AppDatabase): CommentDao = db.commentDao()
}
